app/controllers/security_group_controller.rb

```diff
@@ -12,4 +12,254 @@ class SecurityGroupController < ApplicationController
+    params[:display] = @display if %w(vms instances images).include?(@display)
+    params[:page] = @current_page unless @current_page.nil? # Save current page for list refresh
+
+    @refresh_div = "main_div"
+    return tag("SecurityGroup") if params[:pressed] == "security_group_tag"
+    delete_security_groups if params[:pressed] == 'security_group_delete'
+
+    if params[:pressed] == "security_group_edit"
+      checked_security_group_id = get_checked_security_group_id(params)
+      javascript_redirect :action => "edit", :id => checked_security_group_id
+    elsif params[:pressed] == "security_group_new"
+      javascript_redirect :action => "new"
+    elsif !flash_errors? && @refresh_div == "main_div" && @lastaction == "show_list"
+      replace_gtl_main_div
+    else
+      render_flash
+    end
+  end
+
+  def cancel_action(message)
+    session[:edit] = nil
+    @breadcrumbs.pop if @breadcrumbs
+    javascript_redirect :action    => @lastaction,
+                        :id        => @security_group.id,
+                        :display   => session[:security_group_display],
+                        :flash_msg => message
+  end
+
+  def security_group_form_fields
+    assert_privileges("security_group_edit")
+    security_group = find_by_id_filtered(SecurityGroup, params[:id])
+    render :json => {
+      :name              => security_group.name,
+      :description       => security_group.description,
+      :cloud_tenant_name => security_group.cloud_tenant.name,
+    }
+  end
+
+  def create
+    assert_privileges("security_group_new")
+    case params[:button]
+    when "cancel"
+      javascript_redirect :action    => 'show_list',
+                          :flash_msg => _("Add of new %{model} was cancelled by the user") % {
+                            :model => ui_lookup(:table => 'security_group') }
+
+    when "add"
+      @security_group = SecurityGroup.new
+      options = form_params
+      ems = ExtManagementSystem.find(options[:ems_id])
+      options.delete(:ems_id)
+      task_id = ems.create_security_group_queue(session[:userid], options)
+
+      add_flash(_("Security Group creation failed: Task start failed: ID [%{id}]") %
+                {:id => task_id.inspect}, :error) unless task_id.kind_of?(Fixnum)
+
+      if @flash_array
+        javascript_flash(:spinner_off => true)
+      else
+        initiate_wait_for_task(:task_id => task_id, :action => "create_finished")
+      end
+    end
+  end
+
+  def create_finished
+    task_id = session[:async][:params][:task_id]
+    security_group_name = session[:async][:params][:name]
+    task = MiqTask.find(task_id)
+    if MiqTask.status_ok?(task.status)
+      add_flash(_("%{model} \"%{name}\" created") % {
+        :model => ui_lookup(:table => 'security_group'),
+        :name  => security_group_name
+      })
+    else
+      add_flash(_("Unable to create %{model} \"%{name}\": %{details}") % {
+        :model   => ui_lookup(:table => 'security_group'),
```

Comment:
For the benefit of our translators and for nicer translations, we should better use
`"Unable to create Security Group \"%{name}\": %{details}"`